core/lib/vm_utils/src/vm_env.rs

```diff
@@ -0,0 +1,243 @@
+use anyhow::Context;
+use multivm::{
+    interface::{L1BatchEnv, L2BlockEnv, SystemEnv, TxExecutionMode},
+    vm_latest::constants::BLOCK_GAS_LIMIT,
+};
+use zksync_contracts::BaseSystemContracts;
+use zksync_dal::StorageProcessor;
+use zksync_types::{
+    fee_model::BatchFeeInput, Address, L1BatchNumber, L2ChainId, MiniblockNumber,
+    ProtocolVersionId, H256, ZKPORTER_IS_AVAILABLE,
+};
+
+use crate::db::wait_for_prev_l1_batch_params;
+
+/// The environment required to initialize the VM.
+#[derive(Clone, Debug)]
+pub struct VmEnv {
+    pub l1_batch_env: L1BatchEnv,
+    pub system_env: SystemEnv,
+}
+
+/// The builder for the `VmEnv`. It allows to load the required data from the database.
+#[derive(Clone, Debug)]
+pub struct VmEnvBuilder {
+    miniblock_number: Option<MiniblockNumber>,
+    l1_batch_number: L1BatchNumber,
+    prev_batch_hash: Option<H256>,
+    fee_account: Option<Address>,
+    validation_computational_gas_limit: u32,
+    chain_id: L2ChainId,
+}
+
+impl VmEnvBuilder {
+    pub fn new(
+        l1_batch_number: L1BatchNumber,
+        validation_computational_gas_limit: u32,
+        chain_id: L2ChainId,
+    ) -> Self {
+        Self {
+            l1_batch_number,
+            validation_computational_gas_limit,
+            chain_id,
+            miniblock_number: None,
+            prev_batch_hash: None,
+            fee_account: None,
+        }
+    }
+    pub fn override_miniblock_number(mut self, miniblock_number: MiniblockNumber) -> Self {
+        self.miniblock_number = Some(miniblock_number);
+        self
+    }
+
+    pub fn override_fee_account(mut self, fee_account: Address) -> Self {
+        self.fee_account = Some(fee_account);
+        self
+    }
+
+    /// Builds the `VmEnv` for the executed L1 batch. It verifies that the previous L1 batch is sealed and miniblock
+    /// is included in the L1 batch.
+    pub async fn build_for_executed_batch(
+        mut self,
+        connection: &mut StorageProcessor<'_>,
+    ) -> anyhow::Result<VmEnv> {
+        let batch_hash = connection
+            .blocks_dal()
+            .get_l1_batch_state_root_and_timestamp(self.l1_batch_number - 1)
+            .await?
+            .context(format!(
+                "Previous L1 batch not found {}",
+                self.l1_batch_number - 1
+            ))?
+            .0;
+
+        let (
+            first_miniblock_number_included_in_l1_batch,
+            last_miniblock_number_included_in_l1_batch,
+        ) = connection
+            .blocks_dal()
+            .get_miniblock_range_of_l1_batch(self.l1_batch_number)
+            .await?
+            .context(format!("L1 batch {} not found", self.l1_batch_number))?;
```

Comment:
Nit: [`with_context`](https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.with_context) would be more appropriate here.